src/cmdline_manager.ts

```diff
@@ -116,40 +108,44 @@ export class CommandLineManager implements Disposable {
     }
 
     private cmdlineShow = (content: string, firstc: string, prompt: string, level: number): void => {
-        if (!this.isVisible()) {
-            // Reset the state if this is a new dialog
-            this.reset();
-        }
-
-        this.state.level = level;
-        this.input.title = prompt || this.getTitle(firstc);
         // only redraw if triggered from a known keybinding. Otherwise, delayed nvim cmdline_show could replace fast typing.
         if (!this.state.redrawExpected) {
             logger.debug(`cmdline_show: ignoring cmdline_show because no redraw expected: "${content}"`);
             return;
         }
+
         this.state.redrawExpected = false;
-        this.showInput();
-        if (this.input.value !== content) {
-            logger.debug(`cmdline_show: setting input value: "${content}"`);
-            this.state.lastTypedText = content;
-            this.state.pendingNvimUpdates++;
-            const activeItems = this.input.activeItems; // backup selections
-            this.input.value = content; // update content
-            this.input.activeItems = activeItems; // restore selections
+        const title = prompt || this.getTitle(firstc);
+        const visibilityChanged = this.showInput(level, title, content);
+
+        // We only need to proceed if text has been entered. In both of these cases, they haven't.
+        //
+        // 1. The dialog box is being shown for the first time, and there is no content. No one has typed anything!
+        // 2. The dialog box was already visible, but the content is identical. No one has typed anything!
+        if ((visibilityChanged && content === "") || (!visibilityChanged && this.input.value === content)) {
```

Comment:
why isn't this handled by just `this.input.value === content`? When shown for the first time, this is true since both the value and the content are ""

Reply:
Aha! This is subtle. Check out the log of running `sr'"` both with/without the check for `visibilityChanged`

Without:

```
2024-06-07T13:55:34.854Z CmdLine: cmdline_hide
2024-06-07T13:55:34.854Z CmdLine: visible level is undefined, hiding
2024-06-07T13:55:34.855Z CmdLine: cmdline_hide
2024-06-07T13:55:34.855Z CmdLine: visible level is undefined, hiding
2024-06-07T13:55:36.414Z CmdLine: cmdline_show: "0,call setpos('.', a:tail)"
2024-06-07T13:55:36.414Z CmdLine: dropping cmdline_show as the content is unchanged
2024-06-07T13:55:36.421Z CmdLine: cmdline_hide
2024-06-07T13:55:36.421Z CmdLine: visible level is 1, hiding
2024-06-07T13:55:36.421Z CmdLine: cmdline_show: "0,call setpos('.', a:tail)"
2024-06-07T13:55:36.422Z CmdLine: dropping cmdline_show as the content is unchanged
2024-06-07T13:55:36.424Z CmdLine: cmdline_hide
2024-06-07T13:55:36.424Z CmdLine: visible level is 1, hiding
2024-06-07T13:55:36.428Z CmdLine: onChange: sending cmdline to nvim: "" + "<C-u>call setpos('.', a:tail)" -> "call setpos('.', a:tail)"
2024-06-07T13:55:36.470Z CmdLine: onHide: skipping event
```

With:

```
2024-06-07T13:56:20.232Z CmdLine: cmdline_hide
2024-06-07T13:56:20.232Z CmdLine: visible level is undefined, hiding
2024-06-07T13:56:20.790Z CmdLine: cmdline_show: "0,call setpos('.', a:tail)"
2024-06-07T13:56:20.790Z CmdLine: cmdline_show: setting input value: "call setpos('.', a:tail)", with level 1
2024-06-07T13:56:20.794Z CmdLine: cmdline_hide
2024-06-07T13:56:20.794Z CmdLine: visible level is 1, hiding
2024-06-07T13:56:20.794Z CmdLine: cmdline_show: "0,call setpos('.', a:tail)"
2024-06-07T13:56:20.794Z CmdLine: cmdline_show: setting input value: "call setpos('.', a:tail)", with level 1
2024-06-07T13:56:20.797Z CmdLine: cmdline_hide
2024-06-07T13:56:20.798Z CmdLine: visible level is 1, hiding
2024-06-07T13:56:20.812Z CmdLine: onHide: skipping event
```	

You'll notice that in the second, we don't ignore the input value in this incredibly fast typing case (i.e. send by the plugin). This means that `pendingNvimChanges` isn't incremented, thus causing us to have an onChange event that gets sent to the editor.

Reply:
Reading this comment over, I realize it's kinda unclear what's going on unless you've been in the trenches like I have been. Let me see if I can distill this a bit by speaking event-wise.

In the case `vim-sandwich` provides us, it executes `:setpos('.', a:tail)`. Consider the following events, if the check were just the equality check you propose, just after the extension loads (i.e. the command box has not already been used)

1. The plugin emits `:setpos('.', a:tail)`
2. We drop the event, as the initial content (`:setpos('.', a:tail)`) equals the content we get from nvim (`:setpos('.', a:tail)`). This leaves `pendingNvimEvents at zero`
3. VSCode sees a change from `""` to `":setpos('.', a:tail)"`, so it types this text.
4. We run `sr'"` again. 
5.  We drop the event, as the initial content (`:setpos('.', a:tail)`) equals the content we get from nvim (`:setpos('.', a:tail)`). This leaves `pendingNvimEvents at zero`
6. VSCode does *NOT* see a change (the input box holds `":setpos('.', a:tail)"` and we just set its content to the same), so it does not type to the editor. Thus, the command works fine.

We can also consider the case where we open a new window with the empty string.

1. We type `:`, the box opens with `value` of `""`
2. VSCode will not emit an onChange event, as we went from `""` to `""`
3. We type some text (say, `asdf`), then pres esc.
4. We type `:`, the box opens with a `value` of "".
5. This _does_ fire an onChange event, as we go from `"asdf"` to `""`. However, there's nothing for VSCode to send to the editor here, (as `lastTypedText` has been reset to `""`).

Honestly, after writing all this out, part of me wants to revert to the paradigm where we created a fully new input box for every instance of the command line. However, I have throughly tested these changes and I believe them to be correct. I might do that as a follow-on PR, but the next few days are busy for me so I won't have a chance to commit to it at a level I feel confident about.

Reply:
I found some spare time for the alternate implementation, but I haven't fully tested it. Let me know what you prefer

#2089